File: debop4s-data/src/test/scala/debop4s/data/tests/mapping/tree/TreeNodeTest.scala
package debop4s.data.tests.mapping.tree

import debop4s.core.utils.Hashs
import debop4s.data.jpa.repository.JpaQueryDslDao
import debop4s.data.model.{ LongEntity, HibernateTreeEntity }
import debop4s.data.tests.AbstractJpaTest
import java.util
import javax.persistence._
import org.hibernate.annotations.CacheConcurrencyStrategy
import org.hibernate.{ annotations => hba }
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.transaction.annotation.Transactional

/**
 * Created by debop on 2014. 3. 9.
 */
@Transactional
class TreeNodeTest extends AbstractJpaTest {

  @PersistenceContext val em: EntityManager = null
  @Autowired val dao: JpaQueryDslDao = null

  @Test
  def configurationTest() {
    assert(em != null)
    assert(dao != null)
  }

  @Test
  def buildTree() {
    val root = TreeNode("root")
    val child1 = TreeNode("child1")
    val child2 = TreeNode("child2")

    root.addChild(child1)
    root.addChild(child2)

    val child11 = TreeNode("child11")
    val child12 = TreeNode("child12")
    child1.addChild(child11)
    child1.addChild(child12)

    em.persist(root)
    em.flush()
    em.clear()

    val node = em.find(classOf[TreeNode], child1.id)
    assert(node.children.size == 2)
    assert(node.parent == root)

    // JPQL로 로드하기.
    val roots = em.createQuery("select x from TreeNode x where x.parent is null").getResultList
    assert(roots.size() == 1)
    assert(roots.get(0) == root)
  }

}

@Entity
@org.hibernate.annotations.Cache(region = "tree", usage = CacheConcurrencyStrategy.READ_WRITE)
@hba.DynamicInsert
@hba.DynamicUpdate
@Access(AccessType.FIELD)
class TreeNode extends LongEntity with HibernateTreeEntity[TreeNode] {

  var title: String = _
  var data: String = _
  var description: String = _

  override def self: TreeNode = this

  @ManyToOne(fetch = FetchType.LAZY)
  @hba.LazyToOne(hba.LazyToOneOption.PROXY)
  @JoinColumn(name = "parentId", nullable = true)
  var parent: TreeNode = _

  override def getParent: TreeNode = parent

  override def setParent(parent: TreeNode) {
    this.parent = parent
  }

  @OneToMany(mappedBy = "parent", cascade = Array(CascadeType.ALL), orphanRemoval = true)
  @hba.LazyCollection(hba.LazyCollectionOption.EXTRA)
  override val children: util.Set[TreeNode] = new util.HashSet[TreeNode]()

  @inline
  override def hashCode(): Int = Hashs.compute(title)
}

object TreeNode {

  def apply(title: String, parent: TreeNode = null): TreeNode = {
    val node = new TreeNode()
    node.title = title
    node.parent = parent
    node
  }
}

File: debop4s-data-orm/src/test/java/debop4s/data/orm/jtests/mapping/tree/TreeNodeTest.java
package debop4s.data.orm.jtests.mapping.tree;

import debop4s.data.orm.jtests.jpa.JpaTestBase;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.data.orm.s.mapping.tree.TreeNodeJUnitSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 7. 22. 오후 6:27
 */
@Slf4j
@Transactional
public class TreeNodeTest extends JpaTestBase {

    @PersistenceContext EntityManager em;
    @Autowired TreeNodeRepository repository;

    @Test
    public void buildTreeTest() {
        repository.deleteAllInBatch();
        repository.flush();


        TreeNode root = new TreeNode();
        root.setTitle("root");

        TreeNode child1 = new TreeNode();
        child1.setTitle("child1");

        TreeNode child2 = new TreeNode();
        child1.setTitle("child2");

        root.addChild(child1);
        root.addChild(child2);

        TreeNode child11 = new TreeNode();
        child11.setTitle("child11");

        TreeNode child12 = new TreeNode();
        child12.setTitle("child12");

        child1.addChild(child11);
        child1.addChild(child12);

        repository.save(root);
        repository.flush();
        em.clear();

        TreeNode node = repository.findOne(child1.getId());

        assertThat(node.getChildren().size()).isEqualTo(2);
        assertThat(node.getParent()).isEqualTo(root);


        // JPQL 로 정의된 메소드
        List<TreeNode> roots = repository.findRoots();
        assertThat(roots.size()).isEqualTo(1);
        assertThat(roots.get(0)).isEqualTo(root);

        // 동적 생성
        roots = repository.findByParentIsNull();
        assertThat(roots.size()).isEqualTo(1);
        assertThat(roots.get(0)).isEqualTo(root);

//        List<TreeNode> nodes = Graphs.graphDepthFirstScan(roots.get(0), new Function1<TreeNode, Iterable<TreeNode>>() {
//            @Override
//            public Iterable<TreeNode> execute(TreeNode arg) {
//                return arg.getChildren();
//            }
//        });
//
//        repository.deleteInBatch(nodes);
//        repository.flush();
    }

    @Test
    public void buildTreeAndDelete() throws Exception {
        repository.deleteAllInBatch();
        repository.flush();

        TreeNode root = new TreeNode();
        root.setTitle("root");

        TreeNode child1 = new TreeNode();
        child1.setTitle("child1");

        TreeNode child2 = new TreeNode();
        child1.setTitle("child2");

        root.addChild(child1);
        root.addChild(child2);

        TreeNode child11 = new TreeNode();
        child11.setTitle("child11");

        TreeNode child12 = new TreeNode();
        child12.setTitle("child12");

        child1.addChild(child11);
        child1.addChild(child12);

        repository.saveAndFlush(root);
        em.clear();

        TreeNode child = repository.findOne(child1.getId());
        repository.delete(child);
        em.flush();
        em.clear();

        List<TreeNode> roots = repository.findRoots();
        assertThat(roots).hasSize(1);
        assertThat(roots.get(0).getChildren()).hasSize(1);

    }
}
